### PR TITLE
Move to new Error System ~ Failure and provide new result-based operators

### PR DESCRIPTION
**Pull request by zeeshanlakhani**
Includes:
* new map/transform operators that work over results (so we can handle errors), dropping error'ed packets in a batch and logging that automatically via the error log level. 
  * this is the first hit on this. Group_by and metadata should use a similar approach, as to help us deal with error'ing in NFs (and remove unwraps)
* generally log within NB and let errors be displayed based on Netbricks NF/implementor. 
* error-chain -> failure
* return failure ERROR types as the wrapper, for auto causes and backtraces
* prints become log-levs

### A later todo

@clinedome mentioned wanting to capture any panic!/runtime exception through catch_unwind, but catch_unwind's associated trait does not operate over unsafe code. Once we get past this phase and more cleanup, I'll see what's feasible for a more generic panic-handling solution (i.e. for the panics we don't want to give, e.g. missing config file). 

There will be a related OG PR.

### CODE REVIEW

**Comment by zeeshanlakhani** [comment]
@drunkirishcoder @clinedome def took some opinions here :), but down for any re-doing between this and the "upstream?" pr to OG. I tried to incorporate ideas from previous discussions here by simplifying error'ing and handling batches consistently (at least on transform and map for now). There's still some more work to do, but let's gauge this first before too much is updated.

File: framework/Cargo.toml [hunk 1]
```diff
@@ -8,34 +8,36 @@ build = "build.rs"
 doctest = false
 
 [dependencies]
-libc = ">= 0.2.4"
```
Comment (zeeshanlakhani) [inline framework/Cargo.toml]: most of this is just alphabetical ordering. was tired of the way it was before.

File: framework/src/common/logger.rs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                ),
+                date = chrono::Local::now().format("%Y-%m-%d %H:%M:%S"),
+                target = record.target(),
+                level = colors_level.color(record.level()),
+                message = message,
+            ));
+        })
+        .level(log_level);
+
+    if let Some(e) = ext {
```
Comment (zeeshanlakhani) [inline framework/src/common/logger.rs]: this used for logger extensions, which is pretty minimal right now.

File: framework/src/common/logger.rs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        l = l.level_for(e.nf, e.log_level);
```
Comment (zeeshanlakhani) [inline framework/src/common/logger.rs]: this used for logger extensions, which is pretty minimal right now.

File: framework/src/scheduler/context.rs [hunk 4]
```diff
@@ -211,15 +212,42 @@ impl NetBricksContext {
...
+pub fn initialize_system(configuration: &NetbricksConfiguration) -> Result<NetbricksContext> {
+    init(NetbricksInit {
+        config: configuration,
+        logger_ext: None,
+    })
+}
+
+/// Initialize the system from a more complex setup.
+/// @see init/1
+pub fn initialize_system_2(configuration: NetbricksInit) -> Result<NetbricksContext> {
```
Comment (zeeshanlakhani) [inline framework/src/scheduler/context.rs]: better names always cool :).

File: test/chain-test/src/main.rs [hunk 5]
```diff
@@ -117,9 +117,7 @@ fn main() {
         }
         Err(ref e) => {
             println!("Error: {}", e);
-            if let Some(backtrace) = e.backtrace() {
-                println!("Backtrace: {:?}", backtrace);
-            }
+            println!("Backtrace: {:?}", e.backtrace());
```
Comment (zeeshanlakhani) [inline test/chain-test/src/main.rs]: all these test-case ones below are just b/c backtrace is baked into the `Error` type instead of being an `Option`.

**Comment by drunkirishcoder** [comment]
Still looking through the PR in more details. My first thought is that I think we should follow the rust recommendation that libraries should only depend on `log`, https://github.com/rust-lang-nursery/log#in-libraries. It shouldn't have to know anything about `fern` or any other implementation of `log` and let the application choose what logging framework to use. At least that's one way to go about it, make NetBricks logging implementation agnostic.

**Comment by zeeshanlakhani** [comment]
@drunkirishcoder yeah, that's fine. I kinda liked the NF implementer to not have to care about logging initiation and the like, but moving to the NF Router code is easy.

File: framework/src/scheduler/context.rs [hunk 6]
```diff
@@ -31,9 +32,9 @@ impl<'a> BarrierHandle<'a> {
     }
 }
 
-/// `NetBricksContext` contains handles to all schedulers, and provides mechanisms for coordination.
+/// `NetbricksContext` contains handles to all schedulers, and provides mechanisms for coordination.
 #[derive(Default)]
-pub struct NetBricksContext {
+pub struct NetbricksContext {
```
Comment (drunkirishcoder) [inline framework/src/scheduler/context.rs]: why is this renamed?

File: framework/src/common/errors.rs [hunk 7]
```diff
@@ -1,98 +1,74 @@
...
-    foreign_links {
-        Io(::std::io::Error);
-    }
+use failure::{Error, Fail};
+
+pub type Result<T> = ::std::result::Result<T, Error>;
+
+#[derive(PartialEq, Debug, Fail)]
+
+pub enum NBError {
```
Comment (drunkirishcoder) [inline framework/src/common/errors.rs]: might as well fully spell this out.

File: framework/src/common/errors.rs [hunk 8]
```diff
@@ -1,98 +1,74 @@
...
-                display("Failed to parse MAC address: '{}'", s)
-        }
-    }
-
-    foreign_links {
-        Io(::std::io::Error);
-    }
+use failure::{Error, Fail};
+
+pub type Result<T> = ::std::result::Result<T, Error>;
```
Comment (drunkirishcoder) [inline framework/src/common/errors.rs]: If I'm reading this correctly, this is why we have to do an additional conversion/wrap with `into` everywhere

```
Err(NetBricksError::ConfigurationError("foo).into())
```

I think instead if network functions just bubble up `NetBricksError` instead, we can avoid that.

```
pub type Result<T> = ::std::result::Result<T, NetBricksError>;
```

File: framework/src/common/logger.rs [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    Multiline,
+    SingleLine,
+    SingleLineWithoutBacktrace,
+
+    #[doc(hidden)]
+    __Nonexhaustive,
+}
+
+#[macro_export]
+macro_rules! err {
```
Comment (drunkirishcoder) [inline framework/src/common/logger.rs]: I think all this can be done simpler as a custom implementation of the `Display` and `Debug` traits on `NetBricksError`. Then you can log any error with `error!("{}", e)` or `warn!("{:?}", e)`. Also no confusion that `error!` and `err!` are in fact two different macros but look almost the same.

File: framework/src/config/config_reader.rs [hunk 10]
```diff
@@ -309,14 +316,17 @@ pub fn read_configuration_from_str(
...
 
 /// Read a configuration file and create a `NetbricksConfiguration` structure.
 /// `filename` should be TOML formatted file.
 pub fn read_configuration(filename: &str) -> Result<NetbricksConfiguration> {
     let mut toml_str = String::new();
-    let _ = try! {File::open(filename).and_then(|mut f| f.read_to_string(&mut toml_str))
-    .chain_err(|| ErrorKind::ConfigurationError(String::from("Could not read file")))};
+    let _ = try! {
+        File::open(filename).and_then(|mut f| f.read_to_string(&mut toml_str))
+            .map_err(|e| NBError::ConfigurationError(format!("Could not read file due to error: {}", e)))
```
Comment (drunkirishcoder) [inline framework/src/config/config_reader.rs]: Instead of losing the original error here by converting it to string, we can pass that into the `ConfigurationError` as the `cause` maybe.

File: framework/src/operators/map_results.rs [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+                        err!(e);
+                        self.remove.push(idx)
+                    }
+                }
+            }
+            self.applied = true;
+
+            if !self.remove.is_empty() {
+                self.parent
+                    .drop_packets(&self.remove[..])
```
Comment (drunkirishcoder) [inline framework/src/operators/map_results.rs]: This is great! Love it. I wouldn't mind if we just replace the old `map` and `transform` with this version as the default and only way to do it.

File: framework/src/common/logger.rs (hunk 9 above)
Comment (zeeshanlakhani) [reply framework/src/common/logger.rs]: removed,as backtrace is part of debug calls to `error!` for example.

File: framework/src/operators/map_results.rs (hunk 11 above)
Comment (zeeshanlakhani) [reply framework/src/operators/map_results.rs]: for another time, related to backwards-compat, but maybe just do this as part of the rest of the work toward result operations.

File: framework/src/common/errors.rs (hunk 8 above)
Comment (drunkirishcoder) [reply framework/src/common/errors.rs]: After some discussion with @zeeshanlakhani, we concluded that there's no way to avoid `.into()` calls in general from a network function. Whether we convert everything to `failure::Error` or `NetBricksError` doesn't really make any difference. Keeping it `failure::Error` is likely the simpler design.

**Comment by zeeshanlakhani** [comment]
Test added.

File: framework/src/headers/mac.rs [hunk 12]
```diff
@@ -64,12 +65,11 @@ impl MacAddress {
...
     type Err = Error;
-    fn from_str(s: &str) -> ::std::result::Result<Self, Self::Err> {
+
+    fn from_str(s: &str) -> Result<Self> {
         match hex::decode(s.replace(":", "").replace("-", "")) {
             Ok(ref v) if v.len() == 6 => Ok(MacAddress::new_from_slice(v.as_slice())),
-            _ => Err(Error::from_kind(ErrorKind::FailedToParseMacAddress(
-                s.to_string(),
-            ))),
+            _ => Err(NetBricksError::FailedToParseMacAddress(s.to_string()))?,
```
Comment (drunkirishcoder) [inline framework/src/headers/mac.rs]: We seem to use `into` everywhere else instead of `?` when returning `Err`.

File: framework/src/config/flag_reader.rs [hunk 13]
```diff
@@ -46,17 +43,17 @@ pub fn read_matches(matches: &Matches, opts: &Options) -> NetbricksConfiguration
         match read_configuration(&config_file[..]) {
             Ok(cfg) => cfg,
             Err(ref e) => {
-                print_error(e);
+                error!("{:?}", e);
```
Comment (drunkirishcoder) [inline framework/src/config/flag_reader.rs]: nice. 👍

File: framework/src/common/errors.rs [hunk 14]
```diff
@@ -1,98 +1,76 @@
...
+
+#[derive(Debug, Fail)]
+pub enum NetBricksError {
+    #[fail(display = "Failed to allocate memory")]
+    FailedAllocation,
+
+    #[fail(display = "Failed to deallocate memory")]
+    FailedDeallocation,
+
+    #[fail(display = "Failed to initialize port: {:?}", _0)]
```
Comment (drunkirishcoder) [inline framework/src/common/errors.rs]: `{}` should be sufficient here.

File: framework/src/common/errors.rs [hunk 15]
```diff
@@ -1,98 +1,76 @@
...
+    #[fail(display = "Could not allocate ring")]
+    RingAllocationFailure,
+
+    #[fail(display = "Address of second copy of ring does not match expected address")]
+    RingDuplicationFailure,
+
+    #[fail(display = "Bad ring size {}, must be a power of 2", _0)]
+    InvalidRingSize(usize),
+
+    #[fail(display = "Configuration error: {:?}", _0)]
```
Comment (drunkirishcoder) [inline framework/src/common/errors.rs]: `{}` should be sufficient here.

File: framework/src/common/errors.rs [hunk 16]
```diff
@@ -1,98 +1,76 @@
...
+    RingDuplicationFailure,
+
+    #[fail(display = "Bad ring size {}, must be a power of 2", _0)]
+    InvalidRingSize(usize),
+
+    #[fail(display = "Configuration error: {}", _0)]
+    ConfigurationError(String),
+
+    #[fail(display = "IO Error: {:?}", _0)]
+    IO(#[cause] ::std::io::Error),
```
Comment (drunkirishcoder) [inline framework/src/common/errors.rs]: One last thing. You can delete this.

File: test/transform-error/src/main.rs [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+    WriteLogger::init(
+        LevelFilter::Warn,
+        SimpleConfig {
+            time: None,
+            level: Some(Level::Error),
+            target: Some(Level::Debug),
+            location: Some(Level::Trace),
+            time_format: None,
+        },
+        StdFile::create("test.log").unwrap(),
```
Comment (clinedome-work) [inline test/transform-error/src/main.rs]: so where does this file wind up? should we clean it up?

File: test/transform-error/src/nf.rs [hunk 18]
```diff
@@ -0,0 +1,38 @@
+use failure::Error;
+use netbricks::headers::*;
+use netbricks::operators::*;
+use std::net::Ipv6Addr;
+use std::str::FromStr;
+
+fn throw_mama_from_the_train(hdr: &Ipv6Header) -> Result<(), Error> {
```
Comment (clinedome-work) [inline test/transform-error/src/nf.rs]: 👌

File: framework/src/operators/map_results.rs [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+                    .expect("Filtering bad packet results was performed incorrectly");
```
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply framework/src/operators/map_results.rs]: @clinedome-work yep, exactly. I’ll revisit this call in the next work.

File: test/transform-error/src/main.rs (hunk 17 above)
Comment (zeeshanlakhani) [reply test/transform-error/src/main.rs]: @clinedome-work log files are auto ignored :).

Comment (zeeshanlakhani) [reply test/transform-error/src/main.rs]: Also it’s in the test folder.

File: framework/src/operators/map_results.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (zeeshanlakhani) [reply framework/src/operators/map_results.rs]: Updated this in https://github.com/williamofockham/NetBricks/pull/57/commits/e3c736f2eb3131e3d3f0d25bf18922869fcb0a79 to be smarter and clear out the batch if things get in a bad state with something so simple here.

File: framework/src/common/errors.rs [hunk 20]
```diff
@@ -1,98 +1,145 @@
...
+    #[fail(display = "Failed to parse MAC address: '{}'", _0)]
+    FailedToParseMacAddress(String),
+
+    #[fail(display = "_")]
+    #[doc(hidden)]
+    __Nonexhaustive,
+}
+
+#[macro_export]
+macro_rules! error_chain {
```
Comment (zeeshanlakhani) [inline framework/src/common/errors.rs]: @drunkirishcoder added these macros as the debug print on an `failure::Error` type does not provide the nice display format along with the backtrace, etc... So, to make these useful anyhow, I added to ability to log causes too if you want the entire chain.

File: framework/src/common/errors.rs [hunk 21]
```diff
@@ -71,3 +78,68 @@ pub enum NetBricksError {
     #[doc(hidden)]
     __Nonexhaustive,
 }
+
+#[macro_export]
+macro_rules! error_chain {
+    ($error:expr) => {
+        $crate::common::errors::print_chain($error, log::LevelFilter::Error, true)
+    };
+    () => {};
```
Comment (drunkirishcoder) [inline framework/src/common/errors.rs]: Is this needed? Not sure how I feel about a macro that can resolve to a no op.

File: framework/src/common/errors.rs [hunk 22]
```diff
@@ -71,3 +78,68 @@ pub enum NetBricksError {
...
+    ($error:expr) => {
+        $crate::common::errors::print_chain($error, log::LevelFilter::Warn, false)
+    };
+    () => {};
+}
+
+/// Read a `failure` `Error` and print out the causes and a backtrace as
+/// `log::error`s
+#[doc(hidden)]
+pub fn print_chain(e: &Error, log_level: log::LevelFilter, with_backtrace: bool) {
```
Comment (drunkirishcoder) [inline framework/src/common/errors.rs]: a couple improvements we should make here.

* not write multiple log statements per error. it's ok for console display. but once we start aggregating these across nodes, it will be kinda problematic.
* avoid all this processing if the logging level is turned off. as it is now regardless if logging is on or not, we will loop through the causes and resolve the backtrace.

I think need to restructure the helper a bit to generate the string separate from the logging. something like

```
warn!("{}", stringify(error))
```

File: framework/src/common/errors.rs [hunk 23]
```diff
@@ -71,3 +78,68 @@ pub enum NetBricksError {
...
+        }
+    } else {
+        return ();
+    }
+
+    match (
+        ::std::env::var("RUST_BACKTRACE")
+            .as_ref()
+            .map(|s| s.as_str()),
+        with_backtrace,
```
Comment (drunkirishcoder) [inline framework/src/common/errors.rs]: actually do we need another flag? seems to me it's enough to just key off the env variable. If you want backtrace collected and printed, set env, otherwise don't.

File: framework/src/operators/packet_batch.rs [hunk 24]
```diff
@@ -132,13 +132,15 @@ impl PacketBatch {
                 idx_new += 1;
             }
 
-            // We did not find an index that was passed in, warn/error out.
+            // We did not find an index that was passed in:
+            // Clear the batch
             if remove_idx < idxes.len() {
-                None
+                self.clear_packets();
```
Comment (drunkirishcoder) [inline framework/src/operators/packet_batch.rs]: not sure this is right. if one of the indices passed in is not found we will drop the entire batch?

File: framework/src/operators/packet_batch.rs [hunk 25]
```diff
@@ -147,9 +149,10 @@ impl PacketBatch {
                     let ret = mbuf_free_bulk(array_ptr, len as i32);
                     self.scratch.clear();
                     if ret == 0 {
-                        Some(len)
+                        Ok(len)
                     } else {
-                        None
+                        self.clear_packets();
```
Comment (drunkirishcoder) [inline framework/src/operators/packet_batch.rs]: again not sure what to do, but this seems pretty extreme.

File: framework/src/operators/packet_batch.rs [hunk 26]
```diff
@@ -215,7 +218,7 @@ impl PacketBatch {
                 if ret == 0 {
                     Ok(())
                 } else {
-                    Err(())
+                    Err(NetBricksError::FailedToFreeMBuf.into())
```
Comment (drunkirishcoder) [inline framework/src/operators/packet_batch.rs]: not sure what `mbuf_free_bulk` returns. maybe keep the return code anyway?

File: framework/src/common/errors.rs (hunk 21 above)
Comment (zeeshanlakhani) [reply framework/src/common/errors.rs]: I'll double check, probably can remove.

File: framework/src/operators/packet_batch.rs (hunk 26 above)
Comment (zeeshanlakhani) [reply framework/src/operators/packet_batch.rs]: it returns 0 or 1. but yeah, I can pass that through.

File: framework/src/common/errors.rs (hunk 23 above)
Comment (zeeshanlakhani) [reply framework/src/common/errors.rs]: @drunkirishcoder yeah. I went both ways on that. You don't think we'll want some stracktraces but not always have them "on" in like a prod-case?

File: framework/src/operators/packet_batch.rs (hunk 25 above)
Comment (zeeshanlakhani) [reply framework/src/operators/packet_batch.rs]: after convo, just going to bubble up errors and not clear.

File: framework/src/operators/packet_batch.rs (hunk 26 above)
Comment (zeeshanlakhani) [reply framework/src/operators/packet_batch.rs]: fixing this too. derp

File: framework/src/common/errors.rs [hunk 27]
```diff
@@ -82,64 +81,34 @@ pub enum NetBricksError {
...
-        (Ok("1"), _, log::LevelFilter::Error) => error!("Backtrace:\n{}", e.backtrace()),
-        (Ok("1"), _, log::LevelFilter::Warn) => warn!("Backtrace:\n{}", e.backtrace()),
-        (Ok("0"), true, log::LevelFilter::Error) => error!("Backtrace:\n{}", e.backtrace()),
-        (Ok("0"), true, log::LevelFilter::Warn) => warn!("Backtrace:\n{}", e.backtrace()),
-        _ => (),
+    if let Ok("1") = ::std::env::var("RUST_BACKTRACE")
+        .as_ref()
+        .map(|s| s.as_str())
+    {
+        error.push_str(&format!("Backtrace:\n{}", e.backtrace()))
```
Comment (drunkirishcoder) [inline framework/src/common/errors.rs]: There could be some line breaks quirkiness here I think if there are no `causes`.

File: framework/src/operators/packet_batch.rs (hunk 26 above)
Comment (drunkirishcoder) [reply framework/src/operators/packet_batch.rs]: oh just 0 or one. hmmm. not as useful then. but whatever.
